pkg/endpoint/policy.go

```diff
@@ -155,87 +155,143 @@ func getSecurityIdentities(owner Owner, selector *api.EndpointSelector) []policy
+// removeOldFilter removes the old l4 filter from the endpoint and returns
+// a map of the L3-dependent L4 rules that were successfully removed if the
+// value is true, or in case of an error the value is false.
+func (e *Endpoint) removeOldFilter(owner Owner, filter *policy.L4Filter) (map[policy.NIPortProto]bool, int) {
+	fromEndpointsSrcIDs := map[policy.NIPortProto]bool{}
 	port := uint16(filter.Port)
 	proto, err := u8proto.ParseProtocol(string(filter.Protocol))
 	if err != nil {
 		e.getLogger().WithError(err).Warn("Parse policy protocol failed")
-		return 1
+		return fromEndpointsSrcIDs, 1
 	}
 
 	errors := 0
 	for _, sel := range filter.FromEndpoints {
 		for _, id := range getSecurityIdentities(owner, &sel) {
 			srcID := id.Uint32()
+			nipp := policy.NIPortProto{
+				SecID: id,
+				Port:  port,
+				Proto: uint8(proto),
+			}
 			if err = e.PolicyMap.DeleteL4(srcID, port, uint8(proto)); err != nil {
 				e.getLogger().WithError(err).WithField(logfields.L4PolicyID, srcID).Debug("Delete old l4 policy failed")
+				fromEndpointsSrcIDs[nipp] = false
+			} else {
+				fromEndpointsSrcIDs[nipp] = true
 			}
 		}
 	}
 
-	return errors
+	return fromEndpointsSrcIDs, errors
 }
 
-func (e *Endpoint) applyNewFilter(owner Owner, filter *policy.L4Filter) int {
+// applyNewFilter adds the given l4 filter to the endpoint and returns a map of
+// the L3-dependent L4 rules that were successfully removed if the
+// value is true, or in case of an error the value is false.
+func (e *Endpoint) applyNewFilter(owner Owner, filter *policy.L4Filter) (map[policy.NIPortProto]bool, int) {
+	fromEndpointsSrcIDs := map[policy.NIPortProto]bool{}
 	port := uint16(filter.Port)
 	proto, err := u8proto.ParseProtocol(string(filter.Protocol))
 	if err != nil {
 		e.getLogger().WithError(err).Warn("Parse policy protocol failed")
-		return 1
+		return fromEndpointsSrcIDs, 1
 	}
 
 	errors := 0
 	for _, sel := range filter.FromEndpoints {
 		for _, id := range getSecurityIdentities(owner, &sel) {
 			srcID := id.Uint32()
+			nipp := policy.NIPortProto{
+				SecID: id,
+				Port:  port,
+				Proto: uint8(proto),
+			}
 			if e.PolicyMap.L4Exists(srcID, port, uint8(proto)) {
+				fromEndpointsSrcIDs[nipp] = true
 				e.getLogger().WithField("l4Filter", filter).Debug("L4 filter exists")
 				continue
 			}
 			if err = e.PolicyMap.AllowL4(srcID, port, uint8(proto)); err != nil {
 				e.getLogger().WithError(err).Warn("Update of l4 policy map failed")
 				errors++
+				fromEndpointsSrcIDs[nipp] = false
+			} else {
+				fromEndpointsSrcIDs[nipp] = true
 			}
 		}
 	}
 
-	return errors
+	return fromEndpointsSrcIDs, errors
 }
 
 // Looks for mismatches between 'oldPolicy' and 'newPolicy', and fixes up
 // this Endpoint's BPF PolicyMap to reflect the new L3+L4 combined policy.
-func (e *Endpoint) applyL4PolicyLocked(owner Owner, oldPolicy *policy.L4Policy, newPolicy *policy.L4Policy) error {
+// Returns a map of the L3-dependent L4 rules that were successfully removed and
```

Comment:
Change to "Returns maps of the L3-dependent L4 rules that were successfully removed and added."

 ` if the value of those maps is true, or in case of an error the value is false.` is not clear. Please elaborate.

Reply:
Fixed